xbmc/music/MusicDatabase.cpp

```diff
@@ -7089,53 +7089,114 @@ void CMusicDatabase::SetArtForItem(int mediaId, const std::string &mediaType, co
   }
 }
 
-bool CMusicDatabase::GetArtForItem(int mediaId, const std::string &mediaType, std::map<std::string, std::string> &art)
+bool CMusicDatabase::GetArtForItem(int songId, int albumId, int artistId, bool bPrimaryArtist, std::vector<ArtForThumbLoader> &art)
 {
+  std::string strSQL;
   try
   {
+    if (!(songId > 0 || albumId > 0 || artistId > 0)) return false;
     if (NULL == m_pDB.get()) return false;
     if (NULL == m_pDS2.get()) return false; // using dataset 2 as we're likely called in loops on dataset 1
 
-    std::string sql = PrepareSQL("SELECT type,url FROM art WHERE media_id=%i AND media_type='%s'", mediaId, mediaType.c_str());
-    m_pDS2->query(sql);
+    Filter filter;
+    if (songId > 0)
+      filter.AppendWhere(PrepareSQL("media_id = %i AND media_type ='%s'", songId, MediaTypeSong));
+    if (albumId > 0)
+      filter.AppendWhere(PrepareSQL("media_id = %i AND media_type ='%s'", albumId, MediaTypeAlbum), false);
+    if (artistId > 0)
+      filter.AppendWhere(PrepareSQL("media_id = %i AND media_type ='%s'", artistId, MediaTypeArtist), false);
+
+    strSQL = "SELECT DISTINCT art_id, media_id, media_type, type, '' as prefix, url, 0 as iorder FROM art";
+    if (!BuildSQL(strSQL, filter, strSQL))
+      return false;
+
+    if (!(artistId > 0))
+    {      
+      // Artist ID unknown, so lookup album artist for albums and songs
+      std::string strSQL2;
+      if (albumId > 0)
+      {
+        //Album ID known, so use it to look up album artist(s)
+        strSQL2 = PrepareSQL("SELECT art_id, media_id, media_type, type, "
+          "'albumartist' as prefix, url, album_artist.iOrder as iorder FROM art "
+          "JOIN album_artist ON art.media_id = album_artist.idArtist AND art.media_type ='%s' "
+          "WHERE album_artist.idAlbum = %i ",
+          MediaTypeArtist, albumId);
```

Comment:
Not a biggie, but IMO long statements look better when sql keywords are aligned:

```c++
strSQL2 = PrepareSQL(
  "SELECT art_id, media_id, media_type, type, 'albumartist' as prefix, url, "
  "  album_artist.iOrder as iorder "
  "FROM art "
  "JOIN album_artist "
  "  ON art.media_id = album_artist.idArtist "
  "  AND art.media_type ='%s' "
  "WHERE album_artist.idAlbum = %i ",
  MediaTypeArtist, albumId);
```

Reply:
I particularly dislike that format @garbear, IMO all that white space padding in the SQL string makes for hard to read debug logs, or in other db browsers (for table definitions and views).  Plus people complain MusicDatabase.cpp already has too many lines!
Is 
```
strSQL2 = PrepareSQL("SELECT art_id, media_id, media_type, type, "
            "'albumartist' as prefix, url, album_artist.iOrder as iorder FROM art "
            "JOIN album_artist ON art.media_id = album_artist.idArtist AND art.media_type ='%s' "
            "WHERE album_artist.idAlbum = %i ",
            MediaTypeArtist, albumId);
```
really so hard to read?

Reply:
You make a good point, both are legible. Debug logs aren't an issue because you could readjust the spacing:

```c++
  "SELECT art_id, media_id, media_type, type, 'albumartist' as prefix, url, "
    "album_artist.iOrder as iorder "
```

As for MusicDatabase having too many lines, that's a perfect argument to refactor the lines added here into a new file, so more lines can be used to increase readability.

Reply:
Personally I don't think number of lines is an issue, and splitting SQL across multiple units a bad idea. Actually I would like to move the smartplaylist SQL into the same file as the rest of music db access, have one place where you write SQL queries.

Reply:
Psh. VideoDB is peaking 10000 lines. you got some catching up to do ;)